### DISCUSSION TITLE
Error trying to use generic dataset

### DISCUSSION THREAD

**Discussion by darylsuen** [neuralhydrology/neuralhydrology#184] [General]
I have created a netcdf file with a 'date' coordinate, some forcing variables, and the streamflow target variable using xarray. I tested the netcdf file using the same xarray methods that the genericdataset class method, _load_basin_data(), which implements load_timeseries() uses:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/124c67a9-6bd2-4b97-a6cf-9531fc097f8f)

When my netcdf file opens and I print the dataset I see that the 'date' column is a coordinate and the rest of the columns are variables:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/eeb36f92-2511-49db-8647-4ce7421d76b3)

When I try to run neuralhydrology, I get this error: 

IndexError: index 0 is out of bounds for axis 0 with size 0,

and it is raised because of this code in the basedataset.py:

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/dd0b4cbb-b18b-4f41-96e5-b2c15d789b0f)

I notice that xarray automatically converts day first date objects into year first. I made my 'date' column day first because that's the format used in the config files. If it needs to be day first then the 'date' values have to be converted to strings before  converting to a xarray dataset.  However, this will just raise another error when the native_frequency variable is calculated in the utils.py line 223 which uses pandas.infer_freq() because it expects a DatetimeIndex.

What is a solution?

**Comment by darylsuen** [comment]
The error suggests that the df.index is empty, therefore the dataframe is empty. The dataframe is created with the _load_data() function the BaseDataset class, which implements the _load_or_create_xarray_dataset(), which in turn uses the _load_basin_data() function. I can't seem to find exactly where and when in the program the netcdf file is converted to a dataframe but I think this is creating an empty dataframe. 

Where does this occur so I can check the code and see why an empty dataframe is created?

**Reply by darylsuen** [reply]
The error is not what I at first suspected, I think. I have tried to run the program again and, although I get the same error, the program runs for 3 epochs so the dataframe is not empty.

![image](https://github.com/neuralhydrology/neuralhydrology/assets/115488520/2f7f7c8f-41cc-4e6b-9a8d-f3fab02fb93d)

There are no NaN values in the dataframe and, if there were, no epochs would be completed, anyway. I feel that I am close to getting a generic dataset to work.

**Reply by amywinder900** [reply]
Hi, I don't know if I can help but I can provide a little more information.

I'm using the generic dataset as well and printing my xarray gives the following: 

![image](https://github.com/neuralhydrology/neuralhydrology/assets/89655432/7f1adcfb-87cc-4686-8a31-e9d507175811)

Mine does contains null values so that isn't something that should trip up the code. The obvious differences that I can see are your basin and Id columns. I don't see why that would cause issues but it might be worth deleting them just in case. Other than that, my datetime is in a slightly different format but again I don't see why it would be an issue. I produce mine using pd.datetime. 

Other than that are you using multiple basins or just one? If you use multiple you could trim it down to a single dataset for the moment, as there could be something weird about the data in one file that doesn't exist in the others. 

I hope there's something helpful in there! Good luck.

**Reply by darylsuen** [reply]
Thank amywinder. I managed to get it to work. Still not sure what the issue was. I thought it might be an index issue but the 'Cannot infer a legal frequency...' error message catches those.
